pgjdbc/src/main/java/org/postgresql/jdbc/ArrayDecoding.java

```diff
@@ -299,160 +325,166 @@ public Object parseValue(int length, ByteBuffer bytes, BaseConnection connection
     }
 
     @Override
-    public Object parseValue(String stringVal, BaseConnection connection) throws SQLException {
+    public String parseValue(String stringVal, BaseConnection connection) throws SQLException {
       return stringVal;
     }
+
+    @Override
+    public Class<String> baseType() {
+      return String.class;
+    }
   };
 
-  private static final ArrayDecoder<byte[][]> BYTE_ARRAY_ARRAY = new AbstractObjectArrayDecoder<byte[][]>(
-      byte[].class) {
+  private static final FieldDecoder<byte[]> BYTE_ARRAY = new AbstractObjectFieldDecoder<byte[]>() {
 
     /**
      * {@inheritDoc}
      */
     @Override
-    public Object parseValue(int length, ByteBuffer bytes, BaseConnection connection) throws SQLException {
+    public byte[] parseValue(int length, ByteBuffer bytes, BaseConnection connection) throws SQLException {
       final byte[] array = new byte[length];
       bytes.get(array);
       return array;
     }
 
     @Override
-    public Object parseValue(String stringVal, BaseConnection connection) throws SQLException {
+    public byte[] parseValue(String stringVal, BaseConnection connection) throws SQLException {
       return PGbytea.toBytes(stringVal.getBytes(StandardCharsets.US_ASCII));
     }
+
+    @Override
+    public Class<byte[]> baseType() {
+      return byte[].class;
+    }
   };
 
-  private static final ArrayDecoder<BigDecimal[]> BIG_DECIMAL_STRING_DECODER = new AbstractObjectStringArrayDecoder<BigDecimal[]>(
-      BigDecimal.class) {
+  private static final FieldDecoder<BigDecimal> BIG_DECIMAL_STRING_DECODER = new AbstractObjectFieldDecoder<BigDecimal>() {
 
     @Override
-    public Object parseValue(int length, ByteBuffer bytes, BaseConnection connection) throws SQLException {
+    public BigDecimal parseValue(int length, ByteBuffer bytes, BaseConnection connection) throws SQLException {
       final byte[] numericVal = new byte[length];
       bytes.get(numericVal);
-      return ByteConverter.numeric(numericVal);
+      return (BigDecimal) ByteConverter.numeric(numericVal);
     }
 
     @Override
-    public Object parseValue(String stringVal, BaseConnection connection) throws SQLException {
+    public BigDecimal parseValue(String stringVal, BaseConnection connection) throws SQLException {
       return PgResultSet.toBigDecimal(stringVal);
     }
-  };
-
-  private static final ArrayDecoder<String[]> STRING_ONLY_DECODER = new AbstractObjectStringArrayDecoder<String[]>(
```

Comment:
Hi @bokken, 
I noticed that you have dealt mostly with ArrayDecoding and I wanted to know if there is a reason to keep this decoder since STRING decoder also exists, the only difference is that this one doesn't support binary decoding.
As far as I understand the difference in their usages is that STRING decoder is used for types that are actually text types such as TEXT / VARCHAR / CHAR, while STRING_ONLY_DECODER was used for types that are mapped to PgObjects by us such as json, however they both will return a string at the end so I don't understand why binary decoding would fail for such types.
I tried to remove it in favor of STRING decoder and no tests appear to be failing + types such as json were binary decoded successfully.

Reply:
I do not remember. 
Are you certain that the json (and any other types) were actually being retrieved from the db as binary? Because those oids are not set to support binary by default.

Reply:
Pretty sure JSON is not as the binary format is specific to PostgreSQL

Reply:
I took a look at how the backend deals with the types that were marked as STRING_ONLY before, namely:
bpchar, varchar, char, json, jsonb.

For all of them apart from jsonb I think we can safely add support for binary decoding, as they simply return a string as a byte array.
For reference,
json: https://github.com/postgres/postgres/blob/66188912566b5614dff095ae86f4b1e06d58e875/src/backend/utils/adt/json.c#L137
bpchar: https://github.com/postgres/postgres/blob/66188912566b5614dff095ae86f4b1e06d58e875/src/backend/utils/adt/varchar.c#L251
varchar: https://github.com/postgres/postgres/blob/66188912566b5614dff095ae86f4b1e06d58e875/src/backend/utils/adt/varchar.c#L548
char: https://github.com/postgres/postgres/blob/66188912566b5614dff095ae86f4b1e06d58e875/src/backend/utils/adt/char.c#L105

The exception is with jsonb is that it also sends an extra single byte to represent the version, which I had previously missed,
https://github.com/postgres/postgres/blob/66188912566b5614dff095ae86f4b1e06d58e875/src/backend/utils/adt/jsonb.c#L124

I can add support for this type since is just a matter of skipping past that byte and interpreting the left over part as a string.